src/main/java/net/neoforged/neoforge/transfer/resources/FluidResource.java

```diff
@@ -0,0 +1,298 @@
+     */
+    private final FluidStack innerStack;
+
+    private FluidResource(FluidStack stack) {
+        this.innerStack = stack;
+    }
+
+    @Override
+    public Fluid value() {
+        return innerStack.getFluid();
+    }
+
+    /**
+     * @return The {@link Fluid} of this resource from the inner {@link FluidStack}
+     */
+    public Fluid getFluid() {
+        return value();
+    }
+
+    /**
+     * @return the fluid holder of this resource
+     */
+    @Override
+    public Holder<Fluid> getHolder() {
+        return innerStack.getFluidHolder();
+    }
+
+    /**
+     * @return the fluid type of this resource
+     */
+    public FluidType getFluidType() {
+        return innerStack.getFluidType();
+    }
+
+    /**
+     * Checks if this resource is empty. The resource will be empty if the fluid is {@link Fluids#EMPTY}.
+     *
+     * @return if this resource is empty
+     */
+    @Override
+    public boolean isEmpty() {
+        return innerStack.isEmpty();
+    }
+
+    @Override
+    public FluidResource withPatch(DataComponentPatch patch) {
+        if (isEmpty() || patch.isEmpty() || innerStack.getComponentsPatch().equals(patch))
+            return this;
+
+        FluidStack stack = innerStack.copy();
+        stack.applyComponents(patch);
+        return FluidResource.of(stack);
+    }
+
+    @Override
+    public <D> FluidResource with(DataComponentType<D> type, D data) {
+        if (isEmpty()) return FluidResource.EMPTY;
+        FluidStack stack = innerStack.copy();
+        stack.set(type, data);
+        if (FluidStack.isSameFluidSameComponents(innerStack, stack)) return this;
+        return FluidResource.of(stack);
+    }
+
+    //This is overridden to return FluidResource to allow method chaining
+    @Override
+    public <D> FluidResource with(Supplier<? extends DataComponentType<D>> type, D data) {
+        return with(type.get(), data);
+    }
+
+    @Override
+    public FluidResource without(DataComponentType<?> type) {
+        if (isEmpty()) return FluidResource.EMPTY;
+        FluidStack stack = innerStack.copy();
+        stack.remove(type);
+        if (FluidStack.isSameFluidSameComponents(innerStack, stack)) return this;
+        return FluidResource.of(stack);
+    }
+
+    //This is overridden to return FluidResource to allow method chaining
+    @Override
+    public FluidResource without(Supplier<? extends DataComponentType<?>> type) {
+        return without(type.get());
+    }
+
+    @Override
+    public DataComponentMap getComponents() {
+        return innerStack.immutableComponents();
+    }
+
+    @Override
+    public DataComponentPatch getComponentsPatch() {
+        return innerStack.getComponentsPatch();
+    }
+
+    /**
+     * Creates an {@link FluidStack} of the specified count.
+     *
+     * @param amount The amount of the fluid the stack should have. Must be non-negative.
+     * @return A new copy of the inner fluid stack with the specified count.
```

Comment:
This is an implementation note. The fact that there is an inner stack is irrelevant to the API